controller/volume_controller.go

```diff
@@ -624,16 +623,89 @@ func (vc *VolumeController) cleanupCorruptedOrStaleReplicas(v *longhorn.Volume,
 		// 1. failed before ever became healthy (RW), mostly failed during rebuilding
 		// 2. failed too long ago, became stale and unnecessary to keep
 		// around, unless we don't any healthy replicas
-		// 3. Clean up extra failed to scheduled replica
-		// when no eviction requested.
 		if r.Spec.HealthyAt == "" || (healthyCount != 0 && staled) {
 
-			logrus.Infof("Cleaning up corrupted, staled or extra failed to scheduled replica %v", r.Name)
+			logrus.Infof("Cleaning up corrupted, staled replica %v", r.Name)
 			if err := vc.deleteReplica(r, rs); err != nil {
-				return errors.Wrapf(err, "cannot cleanup the replica %v", r.Name)
+				return errors.Wrapf(err, "cannot cleanup staled replica %v", r.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (vc *VolumeController) cleanupFailedToScheduledReplicas(v *longhorn.Volume, rs map[string]*longhorn.Replica) (err error) {
+	healthyCount := vc.getHealthyReplicaCount(rs)
+	hasEvictionRequestedReplicas := false
+
+	for _, r := range rs {
+		if r.Status.EvictionRequested == true {
+			hasEvictionRequestedReplicas = true
+		}
+	}
+
+	if healthyCount >= v.Spec.NumberOfReplicas {
+		for _, r := range rs {
+			if !hasEvictionRequestedReplicas {
```

Comment:
Hmm? Are we going to have the same issue of repeatedly deleting the local replica in this case?

Reply:
Hi Sheng,

Sorry, I don't fully understand your question. This is for eviction requested volume, in the case of there is no enough resource to evict the replica, we will have one failed to scheduled replica waiting for user ether to add more resource(node) or cancel the eviction. Thanks.

Reply:
I meant, are we going to keep the unscheduled replica due to data locality?